keps/sig-api-machinery/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,281 @@
+in this KEP should be reusable by other components in the future if desired.
+
+### Goals
+
+* Define a clear migration process for existing clusters to use the cloud-controller-manager instead of the kube-controller-manager for cluster integrations with any cloud provider.
+* The migration process should work for single node control planes and HA control planes with minimal risk to users.
+
+### Non-Goals
+
+* Removing cloud provider code from the core Kubernetes tree, this effort is separate and is covered in [KEP-removing-in-tree-providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md)
+* Improving the scalability of controllers by running controllers across multiple components (with or without leader election).
+* Migrating cloud-based volume plugins to CSI. This is a separate effort led by SIG Storage. See [this proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md) for more details.
+
+## Proposal
+
+Today, the kube-controller-manager and cloud-controller-manager run independent of each other.
+This means that both the kube-controller-manager or the cloud-controller-manager can run the
+cloud specific control loops for a given cluster. For a highly available control plane to migrate
+from running only the kube-controller-manager to running both the kube-controller-manager and
+the cloud-controller-manager requires that only one process in the control plane is running
+the cloud specific controllers. This becomes non-trivial when introducing the cloud-controller-manager
+that runs overlapping controllers with the kube-controller-manager.
+
+For environments that can tolerate downtime, it will be recommended to shutdown the control plane,
+reconfigure components to use the cloud-controller-manager, and then bring the control plane back up.
+This ensures that only 1 component can be running the set of cloud controllers. For environments
+that have stricter requirements for uptime, some level of coordination is required between the two
+components to ensure that upgrading control planes does not result in running the same controller
+in more than one place while also accounting for version skew.
+
+In order to coordinate the cloud-specific controllers across the `kube-controller-manager` and the `cloud-controller-manager`,
+this KEP proposes a _primary_ and _secondary_ (a.k.a migration) leader election lock in the `kube-controller-manager`. The primary lock represents
+the current leader election resource lock in the `kube-controller-manager`. The secondary lock would hold the run callback for all
+cloud-specific controllers in the `kube-controller-manager`. The secondary lock would only be run inside the run callback of the primary
+lock. The secondary lock will watch a resource lock named `cloud-controller-manager`. The secondary lock of the `kube-controller-manager`
```

Comment:
If we allow moving (different?) controllers over different migrations then I think we will end up needing different resource lock names. To move the controllers from 1.17 -> 1.18 for a given CP, the lock must be introduced to the KCM in 1.17. It is then removed from the KCM and added to the CCM in 1.18. It has to be in the CCM for all of 1.18 to allow for the possibility of rolling back to 1.17. So it cannot be removed from the CCM till 1.19. However it may be desirable to migrate some other controller from the KCM to the CCM in 1.18 -> 1.19. So I think its safer to call the lock something like "1-17to1-18". Also I believe that "cloud-controller-manager" is already the resource for the primary lock of the cloud-controller-manager.